dlio_benchmark/checkpointing/base_checkpointing.py

```diff
@@ -14,94 +14,319 @@
+        optimizer_checkpoint_size = self.comm.allreduce(optimizer_checkpoint_size)/1024./1024./1024.
+        if self.args.zero_stage < 3:
+            model_checkpoint_size /= self.data_parallelism
+        self.checkpoint_size = model_checkpoint_size + optimizer_checkpoint_size
+        if self.args.my_rank == 0:
+            logging.info(f"{utcnow()} Model size: {model_checkpoint_size:.4f} GB")
+            logging.info(f"{utcnow()} Optimizer state size: {optimizer_checkpoint_size:.4f} GB")
+            logging.info(f"{utcnow()} Total checkpoint size: {self.checkpoint_size:.4f} GB")
 
     @abstractmethod
-    def get_tensor(self, size):
+    def get_tensor(self, length, datatype="int8"):
         return []
 
     @abstractmethod
-    def save_state(self, suffix, state):
+    def save_state(self, suffix, state, fsync=False):
+        pass
+
+    @abstractmethod
+    def load_state(self, suffix, state):
         pass
 
     def get_name(self, suffix):
         return os.path.join(self.args.checkpoint_folder, f"{suffix}.{self.ext}")
 
-    def get_layer_index(self, rank, tensor_parallelism, pipeline_parallelism, total_layers):
-        if tensor_parallelism > 1:
-            total_layers = total_layers + tensor_parallelism
-        
-        divisible_layers = total_layers - (total_layers % pipeline_parallelism)
-        min_layers_per_pipeline = divisible_layers // pipeline_parallelism
-        max_layer_per_pipeline = min_layers_per_pipeline + 1
-        pipeline_rank = (rank // tensor_parallelism) % pipeline_parallelism
-        left_layers = total_layers - divisible_layers
-        num_layers_per_pipeline = max_layer_per_pipeline
-        if pipeline_rank >= left_layers:
-            num_layers_per_pipeline = min_layers_per_pipeline
-        if pipeline_rank < left_layers:
-            start_layer = pipeline_rank * max_layer_per_pipeline
-            end_layer = start_layer + num_layers_per_pipeline - 1
+    def get_num_parameters(self):
+        if self.args.hidden_size <= 0:
+            return 0
+        head_size = self.args.hidden_size//self.args.num_attention_heads
+        dim_kv = head_size * self.args.num_kv_heads        
+        embedding = self.args.vocab_size*self.args.hidden_size
+        input_norm = self.args.hidden_size
+        qkv = self.args.hidden_size * (self.args.hidden_size + 2*dim_kv)
+        dense = self.args.hidden_size*self.args.hidden_size
+        layer_norm = self.args.hidden_size
+        mlp_h_to_4h = self.args.ffn_hidden_size*2*self.args.hidden_size # the factor of 2 is because of gated linear unit                                                                           
+        mlp_4h_to_h = self.args.ffn_hidden_size*self.args.hidden_size
+        weight = self.args.hidden_size
+        lm_head = embedding
+        return embedding  + (input_norm + qkv + dense + layer_norm + mlp_h_to_4h + mlp_4h_to_h)*self.args.num_layers + weight + lm_head
```

Comment:
what is `qkv` and `mlp_h_to_4h`? if `mlp_h_to_4h` is too big then at least add a line comment on what it is, when it is defined.